amaranth/sim/pysim.py

```diff
@@ -314,9 +426,14 @@ def _step(self):
             converged = self._state.commit(changed)
 
         for vcd_writer in self._vcd_writers:
-            for signal_state in changed:
-                vcd_writer.update(self._timeline.now,
-                    signal_state.signal, signal_state.curr)
+            for change in changed:
+                if isinstance(change, _PySignalState):
+                    signal_state = change
+                    vcd_writer.update(self._timeline.now,
+                        signal_state.signal, signal_state.curr)
+                else:
```

Comment:
Please use `elif isinstance(change, _PyMemoryState):`; this lets Pylance propagate type information. (If performance is a concern, using `type(change) is _PyMemoryState` is a much better idea than `isinstance`.)

Reply:
`change` here is a tuple of `_PyMemoryState` and address, though

I'll just add a class for that...